src/utilities/hooks.ts

```diff
@@ -1,27 +1,32 @@
-import { inject, onUnmounted } from 'vue'
-import { useRouterDepth } from '@/compositions/useRouterDepth'
 import { NavigationAbortError } from '@/errors/navigationAbortError'
-import { RouterNotInstalledError } from '@/errors/routerNotInstalledError'
+import { RouteHookStore } from '@/models/RouteHookStore'
 import { RouterPushError, RouterRejectionError } from '@/types/errors'
-import { AddRouteHook, AfterRouteHook, AfterRouteHookResponse, BeforeRouteHook, BeforeRouteHookResponse, RouteHook, RouteHookAbort, RouteHookCondition, RouteHookRemove, RouteHookTiming } from '@/types/hooks'
+import { AfterRouteHook, AfterRouteHookResponse, BeforeRouteHook, BeforeRouteHookResponse, RouteHookAbort, RouteHookLifecycle } from '@/types/hooks'
 import { ResolvedRoute } from '@/types/resolved'
 import { RouterReject } from '@/types/router'
 import { RouterPushImplementation } from '@/types/routerPush'
 import { RouterReplaceImplementation } from '@/types/routerReplace'
-import { asArray } from '@/utilities/array'
-import { RouteHooks, addRouteHookInjectionKey } from '@/utilities/createRouterHooks'
-import { getAfterRouteHooks, getBeforeRouteHooks } from '@/utilities/getRouteHooks'
+import { getAfterRouteHooksFromRoutes, getBeforeRouteHooksFromRoutes } from '@/utilities/getRouteHooks'
 
 type BeforeContext = {
   to: ResolvedRoute,
   from: ResolvedRoute,
-  hooks: RouteHooks,
+  hooks: RouteHookStore,
 }
 
 export async function runBeforeRouteHooks({ to, from, hooks }: BeforeContext): Promise<BeforeRouteHookResponse> {
+  const { global, component } = hooks
+  const route = getBeforeRouteHooksFromRoutes(to, from)
+
   const allHooks: BeforeRouteHook[] = [
-    ...hooks.before,
-    ...getBeforeRouteHooks(to, from),
+    ...global.onBeforeRouteEnter,
+    ...global.onBeforeRouteUpdate,
+    ...global.onBeforeRouteLeave,
+    ...route.onBeforeRouteEnter,
+    ...route.onBeforeRouteUpdate,
+    ...route.onBeforeRouteLeave,
+    ...component.onBeforeRouteUpdate,
+    ...component.onBeforeRouteLeave,
```

Comment:
I had a more complex order in mind something like
```
global before enter
route before enter
global before update
route before update
component before update
component before leave
route before leave
component before leave
```
But once I put them down it seemed to much. The order above seems much more intuitive and explainable. But I don't feel strongly.

Reply:
perhaps we should try extracting this logic so that we can change this easily in the future? looks like we have something similar in `runAfterRouteHooks`

Reply:
despite the way this looks, if a route changes, `global beforeRouteEnter` would be first but `route beforeRouteEnter` would be next right? not `global beforeRouteUpdate`

Reply:
also, are you still planning on unit tests to assert correct ordering?

Reply:
I updated the order to be (before) enter -> update -> leave in  global -> route -> component order and then (after) leave -> update -> enter in component -> route -> global order.